File: src/CircleAI.Observability/InMemoryObservability.cs
// InMemoryObservability.cs
//
// (3.3.0) Real in-memory metric sink, trace sink, and dashboard
// publisher. Metric sink aggregates per-name counters / gauges; trace
// sink stores spans per traceId; dashboard publisher round-trips
// specs by id.

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace CircleAI.Observability;

public sealed class InMemoryMetricSink : IMetricSink
{
    private readonly ConcurrentDictionary<string, List<MetricSample>> _byName = new(StringComparer.Ordinal);
    private readonly object _lock = new();

    public string BackendId => "in-memory";

    public ValueTask EmitAsync(MetricSample sample, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(sample);
        if (string.IsNullOrWhiteSpace(sample.Name)) throw new ArgumentException("Name required");
        lock (_lock)
        {
            var list = _byName.GetOrAdd(sample.Name, _ => new List<MetricSample>());
            list.Add(sample);
        }
        return ValueTask.CompletedTask;
    }

    public IReadOnlyList<MetricSample> Read(string name)
    {
        lock (_lock)
        {
            if (!_byName.TryGetValue(name, out var list)) return Array.Empty<MetricSample>();
            return list.ToArray();
        }
    }

    public IReadOnlyList<string> MetricNames => _byName.Keys.OrderBy(k => k).ToArray();
}

public sealed class InMemoryTraceSink : ITraceSink
{
    private readonly ConcurrentDictionary<string, List<TraceSpan>> _byTrace = new(StringComparer.Ordinal);
    private readonly object _lock = new();

    public string BackendId => "in-memory";

    public ValueTask EmitAsync(TraceSpan span, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(span);
        if (string.IsNullOrWhiteSpace(span.TraceId)) throw new ArgumentException("TraceId required");
        lock (_lock)
        {
            var list = _byTrace.GetOrAdd(span.TraceId, _ => new List<TraceSpan>());
            list.Add(span);
        }
        return ValueTask.CompletedTask;
    }

    public IReadOnlyList<TraceSpan> Read(string traceId)
    {
        lock (_lock)
        {
            if (!_byTrace.TryGetValue(traceId, out var list)) return Array.Empty<TraceSpan>();
            return list.OrderBy(s => s.StartUtc).ToArray();
        }
    }
}

public sealed class InMemoryDashboardPublisher : IDashboardPublisher
{
    private readonly ConcurrentDictionary<string, DashboardSpec> _specs = new(StringComparer.Ordinal);

    public string BackendId => "in-memory";

    public ValueTask PublishAsync(DashboardSpec spec, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(spec);
        if (string.IsNullOrWhiteSpace(spec.DashboardId)) throw new ArgumentException("DashboardId required");
        _specs[spec.DashboardId] = spec;
        return ValueTask.CompletedTask;
    }

    public DashboardSpec? Get(string dashboardId) => _specs.GetValueOrDefault(dashboardId);
    public IReadOnlyList<DashboardSpec> All => _specs.Values.OrderBy(s => s.DashboardId).ToArray();
}

File: python/src/circle_ai/observability/in_memory_observability.py
# in_memory_observability.py
#
# Port of CircleAI.Observability InMemoryObservability.cs (C# — the EXACT spec).
#
# (3.3.0) Real in-memory metric sink (per-name sample lists), trace sink (spans
# per traceId, read ordered by StartUtc), and dashboard publisher (specs by id).
# C# ConcurrentDictionary + lock -> plain dicts guarded by a threading.Lock.

from __future__ import annotations

import threading
from typing import Dict, List, Optional

from .contracts import (
    DashboardSpec,
    IDashboardPublisher,
    IMetricSink,
    ITraceSink,
    MetricSample,
    TraceSpan,
)


class InMemoryMetricSink(IMetricSink):
    """Real in-memory :class:`IMetricSink`. Mirrors
    ``CircleAI.Observability.InMemoryMetricSink``."""

    def __init__(self) -> None:
        self._by_name: Dict[str, List[MetricSample]] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    async def emit_async(
        self, sample: MetricSample, ct: Optional[object] = None
    ) -> None:
        if sample is None:
            raise ValueError("sample")
        if sample.name is None or sample.name.strip() == "":
            raise ValueError("Name required")
        with self._lock:
            self._by_name.setdefault(sample.name, []).append(sample)

    def read(self, name: str) -> List[MetricSample]:
        with self._lock:
            got = self._by_name.get(name)
            return list(got) if got is not None else []

    @property
    def metric_names(self) -> List[str]:
        with self._lock:
            return sorted(self._by_name.keys())


class InMemoryTraceSink(ITraceSink):
    """Real in-memory :class:`ITraceSink`. Mirrors
    ``CircleAI.Observability.InMemoryTraceSink``."""

    def __init__(self) -> None:
        self._by_trace: Dict[str, List[TraceSpan]] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    async def emit_async(
        self, span: TraceSpan, ct: Optional[object] = None
    ) -> None:
        if span is None:
            raise ValueError("span")
        if span.trace_id is None or span.trace_id.strip() == "":
            raise ValueError("TraceId required")
        with self._lock:
            self._by_trace.setdefault(span.trace_id, []).append(span)

    def read(self, trace_id: str) -> List[TraceSpan]:
        with self._lock:
            got = self._by_trace.get(trace_id)
            if got is None:
                return []
            return sorted(got, key=lambda s: s.start_utc)


class InMemoryDashboardPublisher(IDashboardPublisher):
    """Real in-memory :class:`IDashboardPublisher`. Mirrors
    ``CircleAI.Observability.InMemoryDashboardPublisher``."""

    def __init__(self) -> None:
        self._specs: Dict[str, DashboardSpec] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    async def publish_async(
        self, spec: DashboardSpec, ct: Optional[object] = None
    ) -> None:
        if spec is None:
            raise ValueError("spec")
        if spec.dashboard_id is None or spec.dashboard_id.strip() == "":
            raise ValueError("DashboardId required")
        with self._lock:
            self._specs[spec.dashboard_id] = spec

    def get(self, dashboard_id: str) -> Optional[DashboardSpec]:
        with self._lock:
            return self._specs.get(dashboard_id)

    @property
    def all(self) -> List[DashboardSpec]:
        with self._lock:
            return sorted(self._specs.values(), key=lambda s: s.dashboard_id)
